File: go/examples/cost_v2_example.go
package main

import (
	"context"
	"fmt"
	"log"

	cost "github.com/AliyunContainerService/cost-sdk/go"
)

func main() {
	// 创建客户端
	client, err := cost.NewClient(&cost.Config{
		APIServer:  "https://kubernetes.default.svc",
		RetryCount: 3,
		RetryWait:  2,
	})
	if err != nil {
		log.Fatal(err)
	}

	// 示例1: 查询DaemonSet昨天的估算成本明细
	fmt.Println("=== 示例1: 查询DaemonSet昨天的估算成本明细 ===")
	example1(client)

	// 示例2: 获取某个Pod三小时内的估算成本明细，成本集按小时维度拆分
	fmt.Println("\n=== 示例2: 获取某个Pod三小时内的估算成本明细 ===")
	example2(client)

	// 示例3: 获取所有Label Key为app的应用成本，成本集按Label Value进行聚合
	fmt.Println("\n=== 示例3: 按Label聚合应用成本 ===")
	example3(client)
}

func example1(client *cost.Client) {
	// 查询DaemonSet昨天的成本
	response, err := client.CostV2.GetCostV2(context.Background(), &cost.CostV2Request{
		Window: "yesterday",
		Filter: `namespace:"kube-system"+controllerKind:"DaemonSet"+label[app]:"terway-eniip"`,
	})
	if err != nil {
		log.Printf("Error: %v", err)
		return
	}

	for _, data := range response.Data {
		for name, cost := range data {
			fmt.Printf("Name: %s, Cost: %.3f, CPU Request: %.2f, Memory Request: %.0f\n",
				name, cost.Cost, cost.CPUCoreRequestAverage, cost.RAMByteRequestAverage)
		}
	}
}

func example2(client *cost.Client) {
	// 获取某个Pod三小时内的估算成本明细，成本集按小时维度拆分
	response, err := client.CostV2.GetCostV2(context.Background(), &cost.CostV2Request{
		Window: "2024-03-24T00:00:00Z,2024-03-24T03:00:00Z",
		Step:   "1h",
		Filter: `namespace:"kube-system"+pod:"terway-eniip-kz68n"`,
	})
	if err != nil {
		log.Printf("Error: %v", err)
		return
	}

	for _, data := range response.Data {
		for name, cost := range data {
			fmt.Printf("Name: %s, Period: %s - %s, Cost: %.3f\n",
				name, cost.Start, cost.End, cost.Cost)
		}
	}
}

func example3(client *cost.Client) {
	// 获取所有Label Key为app的应用成本，成本集按Label Value进行聚合
	idle := true
	response, err := client.CostV2.GetCostV2(context.Background(), &cost.CostV2Request{
		Window:    "2h",
		Aggregate: "label:app",
		Idle:      &idle,
	})
	if err != nil {
		log.Printf("Error: %v", err)
		return
	}

	for _, data := range response.Data {
		for name, cost := range data {
			fmt.Printf("App: %s, Cost: %.3f, CPU Usage: %.3f, Memory Usage: %.0f\n",
				name, cost.Cost, cost.CPUCoreUsageAverage, cost.RAMByteUsageAverage)
		}
	}
}

File: java/src/main/java/com/aliyun/container/service/cost/sdk/examples/CostV2Example.java
package com.aliyun.container.service.cost.sdk.examples;

import com.aliyun.container.service.cost.sdk.Client;
import com.aliyun.container.service.cost.sdk.Config;
import java.util.Map;
import java.util.HashMap;

public class CostV2Example {
    public static void main(String[] args) {
        try {
            // 创建客户端
            Config config = new Config();
            config.setEndpoint("http://127.0.0.1:8080");
            Client client = new Client(config);

            // 示例1: 查询DaemonSet昨天的估算成本明细
            System.out.println("=== 示例1: 查询DaemonSet昨天的估算成本明细 ===");
            example1(client);

            // 示例2: 获取某个Pod三小时内的估算成本明细，成本集按小时维度拆分
            System.out.println("\n=== 示例2: 获取某个Pod三小时内的估算成本明细 ===");
            example2(client);

            // 示例3: 获取所有Label Key为app的应用成本，成本集按Label Value进行聚合
            System.out.println("\n=== 示例3: 按Label聚合应用成本 ===");
            example3(client);

        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }
    }

    private static void example1(Client client) {
        try {
            // 查询DaemonSet昨天的成本
            Map<String, Object> request = new HashMap<>();
            request.put("window", "yesterday");
            request.put("filter", "namespace:\"kube-system\"+controllerKind:\"DaemonSet\"+label[app]:\"terway-eniip\"");

            Map<String, Object> response = client.getCostV2(request);
            System.out.println("Response code: " + response.get("code"));
            
            if (response.containsKey("data")) {
                System.out.println("Response data: " + response.get("data"));
            }
        } catch (Exception e) {
            System.err.println("Error in example1: " + e.getMessage());
            e.printStackTrace();
        }
    }

    private static void example2(Client client) {
        try {
            // 获取某个Pod三小时内的估算成本明细，成本集按小时维度拆分
            Map<String, Object> request = new HashMap<>();
            request.put("window", "2024-03-24T00:00:00Z,2024-03-24T03:00:00Z");
            request.put("step", "1h");
            request.put("filter", "namespace:\"kube-system\"+pod:\"terway-eniip-kz68n\"");

            Map<String, Object> response = client.getCostV2(request);
            System.out.println("Response code: " + response.get("code"));
            
            if (response.containsKey("data")) {
                System.out.println("Response data: " + response.get("data"));
            }
        } catch (Exception e) {
            System.err.println("Error in example2: " + e.getMessage());
            e.printStackTrace();
        }
    }

    private static void example3(Client client) {
        try {
            // 获取所有Label Key为app的应用成本，成本集按Label Value进行聚合
            Map<String, Object> request = new HashMap<>();
            request.put("window", "2h");
            request.put("aggregate", "label:app");
            request.put("idle", "true");

            Map<String, Object> response = client.getCostV2(request);
            System.out.println("Response code: " + response.get("code"));
            
            if (response.containsKey("data")) {
                System.out.println("Response data: " + response.get("data"));
            }
        } catch (Exception e) {
            System.err.println("Error in example3: " + e.getMessage());
            e.printStackTrace();
        }
    }
}